### PR TITLE
Envoy custom logging format

### PR DESCRIPTION
**Pull request by paulyeo21**
*Issue #, if available:*
N/A

*Description of changes:*
Support envoy access log formatting

* Update VirtualNode CRD
* Update VirtualGateway CRD

*Testing:*
https://quip-amazon.com/19x1AkzcBthF/EKS-Envoy-access-log-custom-formatting-Bug-Bash

Install modified CRDs
```
kubectl apply -k config/helm/appmesh-controller/crds/
```

Grab a virtual node with new custom access logging formatting

```
// virtualnode.yaml

...
logging:
  accessLog:
    file:
      path: "/dev/stdout"
```

```
// virtualnode1.yaml

...
logging:
  accessLog:
    file:
      path: "/dev/stdout"
      format: 
        text: "%LOCAL_REPLY_BODY%:%RESPONSE_CODE%:path=%REQ(:path)%\n"
```

```
// virtualnode2.yaml

...
logging:
  accessLog:
    file:
      path: "/dev/stdout"
      format: 
        json:
          "protocol": "%PROTOCOL%"
          "duration": "%DURATION%"
          "my_custom_header": "%REQ(MY_CUSTOM_HEADER)%"
```

```
kubectl apply -f virtualnode.yaml
kubectl apply -f virtualnode1.yaml
kubectl apply -f virtualnode2.yaml
```

Ensure no validation errors

End to end test:

```
logging:
  accessLog:
    file:
      path: "/dev/stdout"
      format:
        json:
          - key: "protocol"
            value: "jsontestingtesting %PROTOCOL%"
```
Screen Shot 2022-08-04 at 12 37 33

```
logging:
  accessLog:
    file:
      path: "/dev/stdout"
      format:
        text: "texttestingtesting %PROTCOL%:%DURATION%\n"
```
Screen Shot 2022-08-04 at 12 35 58

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: config/crd/bases/appmesh.k8s.aws_virtualgateways.yaml [hunk 1]
```diff
@@ -514,6 +514,33 @@ spec:
...
+                                  properties:
+                                    key:
+                                      description: The name of the field in the JSON
+                                        object
+                                      maxLength: 255
+                                      minLength: 1
+                                      type: string
+                                    value:
+                                      description: The format string
+                                      maxLength: 1024
```
[the remark this answers is not in the document]
Comment (paulyeo21) [reply config/crd/bases/appmesh.k8s.aws_virtualgateways.yaml:532]: these max lengths shouldnt be here actually. let me update

File: go.mod [hunk 2]
```diff
@@ -39,3 +39,5 @@ replace k8s.io/apimachinery => k8s.io/apimachinery v0.21.2
 replace k8s.io/cli-runtime => k8s.io/cli-runtime v0.21.2
 
 replace github.com/containerd/containerd => github.com/containerd/containerd v1.5.13
+
+replace github.com/aws/aws-sdk-go => ./vendor/github.com/aws/aws-sdk-go
```
Comment (paulyeo21) [inline go.mod:43]: this will be removed before merge

File: pkg/conversions/virtualnode_types_conversion.go [hunk 3]
```diff
@@ -589,8 +589,42 @@ func Convert_CRD_ServiceDiscovery_To_SDK_ServiceDiscovery(crdObj *appmesh.Servic
 	return nil
 }
 
+func Convert_CRD_JsonFormatRef_To_SDK_JsonFormatRef(crdObj *appmesh.JsonFormatRef, sdkObj *appmeshsdk.JsonFormatRef) error {
```
Comment (joesbigidea) [inline pkg/conversions/virtualnode_types_conversion.go:592]: why is this declared to return error?

**Review by joesbigidea — COMMENTED** [review]
Needs unit tests around conversion, it would be good to add checks on formats into integration tests as well, once the API is ready

File: pkg/conversions/virtualnode_types_conversion.go (hunk 3 above)
Comment (paulyeo21) [reply pkg/conversions/virtualnode_types_conversion.go:592]: yeah true we can remove it; i think i was following the existing functions like here https://github.com/aws/aws-app-mesh-controller-for-k8s/pull/609/files#diff-04df0bfad47aef04c49cc263ca8b33cde2f9c2016a876a2ecee4a704397f6e0dR542

**Comment by paulyeo21** [comment]
> Needs unit tests around conversion, it would be good to add checks on formats into integration tests as well, once the API is ready

sounds good; ill get the unit tests in now and the integration test later

File: pkg/conversions/virtualgateway_types_conversion_test.go [hunk 4]
```diff
@@ -1163,40 +1163,46 @@ func TestConvert_CRD_VirtualGatewayListenerTLS_To_SDK_VirtualGatewayListenerTLS(
 }
 
 func TestConvert_CRD_VirtualGatewayFileAccessLog_To_SDK_VirtualGatewayFileAccessLog(t *testing.T) {
-	type args struct {
-		crdObj *appmesh.VirtualGatewayFileAccessLog
-		sdkObj *appmeshsdk.VirtualGatewayFileAccessLog
-		scope  conversion.Scope
-	}
```
Comment (paulyeo21) [inline pkg/conversions/virtualgateway_types_conversion_test.go:1170]: this isn't essential but i cleaned this up bc its unnecessarily verbose imo

File: pkg/conversions/virtualnode_types_conversion_test.go [hunk 5]
```diff
@@ -3790,3 +3790,68 @@ func TestConvert_CRD_VirtualNodeSpec_To_SDK_VirtualNodeSpec(t *testing.T) {
 		})
 	}
 }
+
+func newStringPtr(s string) *string {
+	return &s
+}
```
Comment (paulyeo21) [inline pkg/conversions/virtualnode_types_conversion_test.go:3796]: need this to make one liner string ptrs in Go (sucks!)

Comment (jeremymill) [reply pkg/conversions/virtualnode_types_conversion_test.go:3796]: I think we typically use aws.String() for this. https://docs.aws.amazon.com/sdk-for-go/api/aws/

Comment (paulyeo21) [reply pkg/conversions/virtualnode_types_conversion_test.go:3796]: ah yes that works, thanks

**Comment by joesbigidea** [comment]
Code looks good, looks like tests/integration-tests are mucked up

**Comment by paulyeo21** [comment]
> Code looks good, looks like tests/integration-tests are mucked up

yeah its going to fail to compile because its pointing to a vendor directory which we dont check-in. it works locally though. when the SDK is GA I can update this PR and it should be able to compile then

File: go.mod (hunk 2 above)
Comment (ar4096) [inline go.mod:43]: Is this temporary ?

**Review by joesbigidea — APPROVED** [review]

**Pull request merged**